lumicks/pylake/file_download.py

```diff
@@ -0,0 +1,139 @@
+    response = requests.get(url)
+
+    if not response.ok:
+        raise RuntimeError(f"DOI could not be resolved {response}")
+
+    return response.url
+
+
+def download_record_metadata(record_number):
+    """Download specific Zenodo record metadata"""
+    zenodo_url = "https://zenodo.org/api/records/"
+    response = requests.get(urljoin(zenodo_url, str(record_number)))
+    if response.ok:
+        return response.json()
+    else:
+        raise ValueError(f"Failed to access Zenodo record {response}")
+
+
+def download_file(url, target_path, file_name, show_progress=True, block_size=8192):
+    """Stream a file from a URL while showing progress
+
+    Parameters
+    ----------
+    url : str
+        URL to pull from
+    target_path : str
+        Target path to download to
+    file_name : str
+        Filename to download
+    show_progress : bool
+        Show progress while downloading or not?
+    block_size : int
+        Block size to use when downloading
+    """
+    response = requests.get(url, stream=True)
+    size_bytes = int(response.headers.get("content-length", 0))
+
+    full_path = os.path.join(target_path, file_name)
+    dir_name = os.path.dirname(full_path)
+    if dir_name:
+        os.makedirs(dir_name, exist_ok=True)
+
+    with open(full_path, "wb") as file:
+        with tqdm(
+            total=size_bytes,
+            unit="iB",
+            unit_scale=True,
+            desc=f"Download {file_name}",
+            disable=not show_progress,
+        ) as progress_bar:
+            for data in response.iter_content(block_size):
+                progress_bar.update(len(data))
+                file.write(data)
+
+
+def verify_hash(file_name, algorithm, reference_hash, chunk_size=65536):
+    """Verify the hash of a file"""
+    m = hashlib.new(algorithm)
+    with open(file_name, "rb") as f:
+        b = f.read(chunk_size)
+        while len(b) > 0:
+            m.update(b)
+            b = f.read(chunk_size)
+
+    return m.hexdigest() == reference_hash
+
+
+def download_from_doi(doi, target_path="", force_download=False, show_progress=True):
+    """Download files from a Zenodo DOI (i.e. 10.5281/zenodo.#######) and returns a file list in the
+    form of a List[str].
+
+    Parameters
+    ----------
+    doi : str
+        DOI of the files to download.
+    target_path : str
+        Target path to download to. Default downloads to current folder.
+    force_download : bool
+        Force re-downloading the file even if it already exists and the hash checks out.
+        When the hash of an existing file does not match, this will overwrite the existing file with
+        a freshly downloaded copy.
+    show_progress : bool
+        Show a progress bar while downloading.
+    """
+    url = get_url_from_doi(doi)
+
+    if "zenodo" not in url:
+        raise RuntimeError("Only Zenodo DOIs are supported.")
+
+    record_number = url.split("/")[-1].strip()
+    rec = download_record_metadata(record_number)
```

Comment:
would prefer `record_meta` instead of `rec`

Reply:
👍